Feature: Add poetry as pre-commit hook

Cielquan opened the pull request:
# Pull Request Check List

Resolves: #2457 

<!-- This is just a reminder about the most common mistakes. Please make sure that you tick all *appropriate* boxes.  But please read our [contribution guide](https://python-poetry.org/docs/contributing/) at least once, it will save you unnecessary review cycles! -->

- [x] Added **tests** for changed code. (no code was changed)
- [x] Updated **documentation** for changed code.

<!-- **Note**: If your Pull Request introduces a new feature or changes the current behavior, it should be based
on the `develop` branch. If it's a bug fix or only a documentation update, it should be based on the `master` branch. -->

<!-- If you have *any* questions to *any* of the points above, just **submit and ask**!  This checklist is here to *help* you, not to deter you from contributing! -->

I created this draft PR for easier discussion in the issue #2457 .

SanketDG commented on .pre-commit-hooks.yaml:
```diff
@@ -0,0 +1,22 @@
+  language: python
+  language_version: python3
+  pass_filenames: false
+- id: poetry-lock
```
could you add a newline after each hook?

Cielquan replied:
Sure and done.

SanketDG commented on docs/docs/cli.md:
```diff
@@ -385,6 +385,11 @@ poetry shell
 
+!!!note
+
+    This command is also available as `pre-commit` hook.
```
```suggestion
    This command is also available as a `pre-commit` hook.
```

Cielquan replied:
Changed at all 3 occurrences.

SanketDG commented on .pre-commit-hooks.yaml:
```diff
@@ -0,0 +1,22 @@
+  language: python
+  language_version: python3
+  pass_filenames: false
+  args: ["-f", "requirements.txt"]
```
I thinka adding a `-o requirements.txt` makes sense here as default behaviour

Cielquan replied:
Yeah I think you are right.
My initial thought was that the command does not fail my default. But yeah I think `-o requirements.txt` will be my far the most usecase.

Changed it.

Cielquan replied:
Updated docs accordingly.

SanketDG commented on docs/docs/pre-commit-hooks.md:
```diff
@@ -0,0 +1,72 @@
+
+### Arguments
+
+The hook takes the same arguments as the CLI command.
```
Replace `CLI` with `poetry`?

Cielquan replied:
Yeah sound better.

Changed it.

Cielquan commented:
Thanks for the review. I fixed all issues.

But before I make the draft final I want to get feedback on the topic with the chicken-egg-problem I mentioned in the [comments of the connected issue](https://github.com/python-poetry/poetry/issues/2457#issuecomment-640047433).

I fixed the problem with the first solution (bootstrapping) in commit https://github.com/python-poetry/poetry/pull/2511/commits/1121d06a52aceecae7edfdd9d12bee50b15bc80a but I would like to hear others opinions first.

Cielquan commented:
#2515 is solved on develop and will be released with 1.1 so I will change this PR to use develop.

Cielquan commented:
Rebased the branch onto develop and changed merge target to develop.

Cielquan commented:
@sdispater Do you have any idea why the `Tests / FreeBSD / PYTHON:python3.5` pipeline fails?

I found the same issue in your PR #2509 and the PR #2448 .

The similarity is that all 3 have develop as target. But there are other PRs with target develop which succeeded in between.

```
py run-test: commands[1] | poetry run pytest -q --junitxml=junit.xml tests tests/
Traceback (most recent call last):
  File "/tmp/cirrus-ci-build/.tox/py/bin/poetry", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/tmp/cirrus-ci-build/.tox/py/lib/python3.5/site-packages/pkg_resources/__init__.py", line 3252, in <module>
    @_call_aside
  File "/tmp/cirrus-ci-build/.tox/py/lib/python3.5/site-packages/pkg_resources/__init__.py", line 3236, in _call_aside
    f(*args, **kwargs)
  File "/tmp/cirrus-ci-build/.tox/py/lib/python3.5/site-packages/pkg_resources/__init__.py", line 3265, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/tmp/cirrus-ci-build/.tox/py/lib/python3.5/site-packages/pkg_resources/__init__.py", line 584, in _build_master
    ws.require(__requires__)
  File "/tmp/cirrus-ci-build/.tox/py/lib/python3.5/site-packages/pkg_resources/__init__.py", line 901, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/tmp/cirrus-ci-build/.tox/py/lib/python3.5/site-packages/pkg_resources/__init__.py", line 787, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'typing-extensions<4.0,>=3.6; python_version >= "3.5.0" and python_version < "3.5.4"' distribution was not found and is required by clikit
ERROR: InvocationError for command /tmp/cirrus-ci-build/.tox/py/bin/poetry run pytest -q --junitxml=junit.xml tests tests/ (exited with code 1)
___________________________________ summary ____________________________________
ERROR:   py: commands failed
```

abn commented:
@Cielquan there is a bug in clikit that is being resolved. Should be fixed soon.

https://github.com/sdispater/clikit/pull/32

abn commented:
@Cielquan please rebase this to trigger CI with the fix.

Cielquan commented:
Problems solved themselves \o/ ? :fireworks:

Diaoul commented:
You cannot have additional arguments like `--extras` or `--without-hashes` without overriding the whole `args`.
So it's not exactly the same feature-set as suggested in my implementation at #2457 where I had to wrap the current CLI in a more pre-commit friendly [script](https://github.com/Diaoul/pre-commit-hooks-poetry/blob/master/pre_commit_hooks/export.py), take a look at the [README](https://github.com/Diaoul/pre-commit-hooks-poetry/blob/master/README.md).

I don't remember why I didn't use the `pass_filenames: false` approach, maybe it had something to do with how pre-commit reports an error if you don't specify it.

I guess this is fine to leave out some complexity for the user to handle for the sake of a simpler implementation in the code.

Diaoul commented:
> I don't remember why I didn't use the pass_filenames: false approach, maybe it had something to do with how pre-commit reports an error if you don't specify it.

I just tested this, errors are reported the same way, so I think the reason probably was to use `files` instead of using `args` to specify the file. So you could just do that without messing with the whole args.

Cielquan commented:
> You cannot have additional arguments like `--extras` or `--without-hashes` without overriding the whole `args`.
I am fully aware of this. But I did not clarify that in the docs.
- [ ] I will update them.

> So it's not exactly the same feature-set as suggested in my implementation at #2457 where I had to wrap the current CLI in a more pre-commit friendly [script](https://github.com/Diaoul/pre-commit-hooks-poetry/blob/master/pre_commit_hooks/export.py), take a look at the [README](https://github.com/Diaoul/pre-commit-hooks-poetry/blob/master/README.md).
> 
> I don't remember why I didn't use the `pass_filenames: false` approach, maybe it had something to do with how pre-commit reports an error if you don't specify it.
> 
> I guess this is fine to leave out some complexity for the user to handle for the sake of a simpler implementation in the code.

Well I use my branch with pre-commit like so:
```yaml
  - repo: https://github.com/Cielquan/poetry
    rev: 4c3338d18d319fb7edea02d653d2c65622d84ff2
    hooks:
      #: Check config file
      - id: poetry-check
      #: Update lock file
      - id: poetry-lock
```
Those two hooks did work till now without a problem, but they also don't take args.

When I wrote the branch I tested it manually and did not see any errors.

___

> > I don't remember why I didn't use the pass_filenames: false approach, maybe it had something to do with how pre-commit reports an error if you don't specify it.
> 
> I just tested this, errors are reported the same way, so I think the reason probably was to use `files` instead of using `args` to specify the file. So you could just do that without messing with the whole args.

Well I understand your reasoning but normally you write your `.pre-commit-config.yaml` once and don't update it every commit. So IMO its fine if the user has to specify the default args *again* if he/she wants to change the behavior. I also think its more clean for the enduser this way. With the updated docs this should be fine.

Cielquan commented:
@Diaoul 

> Cielquan dismissed Diaoul’s stale review via 43b0fd8 13 minutes ago

While you suggested a change I was changing the docs and overwrote it without noticing.

Diaoul commented:
Did you run into issues installing pre-commit with poetry? On my computer it's pulling weird dependencies, I need to reproduce on a clean environment to see if I can confirm that behavior.

Cielquan commented:
I did not have any issues till now.

Diaoul commented:
Weird, on my computer it tries to install a bunch of stuff and fails on mysql-connector-python-rf.
pip doesn't even install that.

```
pre-commit 2.5.1 A framework for managing and maintaining multi-language pre-commit hooks.
├── cfgv >=2.0.0
├── identify >=1.0.0
├── importlib-metadata *
│   └── zipp >=0.5 
├── importlib-resources *
│   ├── importlib-metadata * 
│   │   └── zipp >=0.5 
│   └── zipp >=0.4 (circular dependency aborted here)
├── nodeenv >=0.11.1
├── pyyaml >=5.1
├── toml *
└── virtualenv >=20.0.8
    ├── dnspython * 
    ├── mysnow * 
    │   ├── mysql-connector-python-rf * 
    │   ├── requests * 
    │   │   ├── certifi >=2017.4.17 
    │   │   ├── chardet >=3.0.2,<4 
    │   │   ├── idna >=2.5,<3 
    │   │   └── urllib3 >=1.21.1,<1.25.0 || >1.25.0,<1.25.1 || >1.25.1,<1.26 
    │   └── wheel * 
    ├── mysql-connector-python-rf * (circular dependency aborted here)
    ├── paramiko * 
    │   ├── bcrypt >=3.1.3 
    │   │   ├── cffi >=1.1 
    │   │   │   └── pycparser * 
    │   │   └── six >=1.4.1 
    │   ├── cryptography >=2.5 
    │   │   ├── cffi >=1.8,<1.11.3 || >1.11.3 (circular dependency aborted here)
    │   │   └── six >=1.4.1 (circular dependency aborted here)
    │   └── pynacl >=1.0.1 
    │       ├── cffi >=1.4.1 (circular dependency aborted here)
    │       └── six * (circular dependency aborted here)
    ├── requests * (circular dependency aborted here)
    ├── setuptools * 
    └── wheel * (circular dependency aborted here)
```

Unrelated to this PR but I think if this is an actual issue it's pretty embarrassing 😉 .

Cielquan commented:
Yeah thats off-topic. I have an answer for you. If you want to discuss this further you are welcome to write me an email (find it at my gh profile).

<details> 
  <summary>My answer</summary>
   That's my tree ... the sub-dependencies from your virtualenv are plenty more than mine. In my venv `virtualenv==20.0.23` is installed same as in your pip output.
    Currently I work with this: `pre-commit = "^2.4"`
    
    ```console
    $ poetry show pre-commit --tree
    pre-commit 2.5.1 A framework for managing and maintaining multi-language pre-commit hooks.
    ├── cfgv >=2.0.0
    ├── identify >=1.0.0
    ├── importlib-metadata *
    │   └── zipp >=0.5 
    ├── importlib-resources *
    │   ├── importlib-metadata * 
    │   │   └── zipp >=0.5 
    │   └── zipp >=0.4 (circular dependency aborted here)
    ├── nodeenv >=0.11.1
    ├── pyyaml >=5.1
    ├── toml *
    └── virtualenv >=20.0.8
        ├── appdirs >=1.4.3,<2 
        ├── distlib >=0.3.0,<1 
        ├── filelock >=3.0.0,<4 
        ├── importlib-metadata >=0.12,<2 
        │   └── zipp >=0.5 
        ├── importlib-resources >=1.0 
        │   ├── importlib-metadata * (circular dependency aborted here)
        │   └── zipp >=0.4 (circular dependency aborted here)
        └── six >=1.9.0,<2 
    ```
</details>

abn commented:
We no longer use the `develop` branch for development PRs. Please rebase on `master`.

Cielquan commented:
> We no longer use the `develop` branch for development PRs. Please rebase on `master`.

Rebased successfully onto master. Thanks for telling.

Cielquan commented:
Rebased onto current master to trigger py39 CI tests

Cielquan commented:
@python-poetry/triage
I would like to kindly ask if the PR is ok like this? Because its open some months now and I would love to see this enhancement in poetry :smiley:

sinoroc commented:
Personal opinion: Shouldn't it be a project of its own?

Diaoul commented:
> Personal opinion: Shouldn't it be a project of its own?

I think it is common practice to host the hooks on the main repo itself: [Black](https://github.com/psf/black/blob/master/.pre-commit-hooks.yaml) does this and so does [flake8](https://gitlab.com/pycqa/flake8/-/blob/master/.pre-commit-hooks.yaml)

Cielquan commented:
> > Personal opinion: Shouldn't it be a project of its own?
> 
> I think it is common practice to host the hooks on the main repo itself: [Black](https://github.com/psf/black/blob/master/.pre-commit-hooks.yaml) does this and so does [flake8](https://gitlab.com/pycqa/flake8/-/blob/master/.pre-commit-hooks.yaml)

I see it like @Diaoul.

I think its better if the `.pre-commit-hooks.yaml` file is in the original project instead of a mirror, because if you want to use the project with pre-commit you can see it right away that its supported and you don't have to search for a mirror and so on.
And after all its only a config file which does no harm to the project itself.

vgel commented:
Can you add a hook to also export dev dependencies? (`poetry export --dev -f requirements.txt -o requirements-test.txt`). This is useful for CI environments like [Heroku](https://devcenter.heroku.com/articles/heroku-ci#python).

```yaml
- id: poetry-export-dev
  name: poetry-export-dev
  description: run poetry export --dev to sync lock file with requirements-test.txt
  entry: poetry export
  language: python
  language_version: python3
  pass_filenames: false
  args: ["--dev", "-f", "requirements.txt", "-o", "requirements-test.txt"]
```

vgel commented:
Also in [the poetry Heroku buildpack](https://github.com/StriveTalent/python-poetry-buildpack/blob/5b9d3e800467857d7d2670d2c2b0f66cbde2eb0e/bin/compile#L44-L50)[1] and in [this comment](https://github.com/python-poetry/poetry/issues/2457#issuecomment-713131930), the file is exported with the `--without-hashes` option. This appears to be due to this issue: https://github.com/pypa/pip/issues/4995, which relates to editable dependencies. It seems to me that since the bug where Poetry was exporting local dependencies as editable has been fixed, the `--without-hashes` argument shouldn't be needed, but wanted to confirm that here?

1: This is our fork of the poetry buildpack which also adds the dev dependencies export.

Cielquan commented:
> Can you add a hook to also export dev dependencies? (`poetry export --dev -f requirements.txt -o requirements-test.txt`). This is useful for CI environments like [Heroku](https://devcenter.heroku.com/articles/heroku-ci#python).
> 
> ```yaml
> - id: poetry-export-dev
>   name: poetry-export-dev
>   description: run poetry export --dev to sync lock file with requirements-test.txt
>   entry: poetry export
>   language: python
>   language_version: python3
>   pass_filenames: false
>   args: ["--dev", "-f", "requirements.txt", "-o", "requirements-test.txt"]
> ```
> 
> (I'm not sure if "dev" -> "test" is the ideal mapping, but poetry doesn't have a "test" section, and things like pytest are usually installed under dev in my experience)

I would not like to do that as it would be only doubling the existing `poetry-export` hook with slightly other default `args`.

I don't know how well you know `pre-commit` but the values in `.pre-commit-hooks.yaml` are only default values which you can overwrite/customize in your own `.pre-commitconfig.yaml`.
So my recommended and preferred way of solving this would be to customize the hook in your `.pre-commit-config.yaml` like so:
```yaml
repos:
  - repo: https://github.com/python-poetry/poetry
    rev: <branch name / tag / sha>
    hooks:
      - id: poetry-export
        name: poetry-export-dev
        args: ["--dev", "-f", "requirements.txt", "-o", "requirements.txt"]
```
For my example you have to add a corresponding `rev` value after this PR is merged.
You just need to overwrite the default `args` and optionally you can overwrite the `name`, which is shown when `pre-commit` runs, too.

_______________


> Also in [the poetry Heroku buildpack](https://github.com/StriveTalent/python-poetry-buildpack/blob/5b9d3e800467857d7d2670d2c2b0f66cbde2eb0e/bin/compile#L44-L50)[1] and in [this comment](https://github.com/python-poetry/poetry/issues/2457#issuecomment-713131930), the file is exported with the `--without-hashes` option. This appears to be due to this issue: [pypa/pip#4995](https://github.com/pypa/pip/issues/4995), which relates to editable dependencies. It seems to me that since the bug where Poetry was exporting local dependencies as editable has been fixed, the `--without-hashes` argument shouldn't be needed, but wanted to confirm that here?
> 
> 1: This is our fork of the poetry buildpack which also adds the dev dependencies export.

I hear about this bug for the first time, but if it still persists I could and would like to add `--without-hashes` as to the default `args`.

vgel commented:
I didn't know about the argument override, that's cool! In that case it
shouldn't be necessary to add the dev dependencies feature (though it's
maybe worth documenting).

++++1 to this PR from me :)

On Thu, Nov 5, 2020 at 12:40 AM Christian Riedel <notifications@github.com>
wrote:

> Can you add a hook to also export dev dependencies? (poetry export --dev
> -f requirements.txt -o requirements-test.txt). This is useful for CI
> environments like Heroku
> <https://devcenter.heroku.com/articles/heroku-ci#python>.
>
> - id: poetry-export-dev
>   name: poetry-export-dev
>   description: run poetry export --dev to sync lock file with requirements-test.txt
>   entry: poetry export
>   language: python
>   language_version: python3
>   pass_filenames: false
>   args: ["--dev", "-f", "requirements.txt", "-o", "requirements-test.txt"]
>
> (I'm not sure if "dev" -> "test" is the ideal mapping, but poetry doesn't
> have a "test" section, and things like pytest are usually installed under
> dev in my experience)
>
> I would not like to do that as it would be only doubling the existing
> poetry-export hook with slightly other default args.
>
> I don't know how well you know pre-commit but the values in
> .pre-commit-hooks.yaml are only default values which you can
> overwrite/customize in your own .pre-commitconfig.yaml.
> So my recommended and preferred way of solving this would be to customize
> the hook in your .pre-commit-config.yaml like so:
>
> repos:
>   - repo: https://github.com/python-poetry/poetry
>     rev: <branch name / tag / sha>
>     hooks:
>       - id: poetry-export
>         name: poetry-export-dev
>         args: ["--dev", "-f", "requirements.txt", "-o", "requirements.txt"]
>
> For my example you have to add a corresponding rev value after this PR is
> merged.
> You just need to overwrite the default args and optionally you can
> overwrite the name, which is shown when pre-commit runs, too.
> ------------------------------
>
> Also in the poetry Heroku buildpack
> <https://github.com/StriveTalent/python-poetry-buildpack/blob/5b9d3e800467857d7d2670d2c2b0f66cbde2eb0e/bin/compile#L44-L50>[1]
> and in this comment
> <https://github.com/python-poetry/poetry/issues/2457#issuecomment-713131930>,
> the file is exported with the --without-hashes option. This appears to be
> due to this issue: pypa/pip#4995 <https://github.com/pypa/pip/issues/4995>,
> which relates to editable dependencies. It seems to me that since the bug
> where Poetry was exporting local dependencies as editable has been fixed,
> the --without-hashes argument shouldn't be needed, but wanted to confirm
> that here?
>
> 1: This is our fork of the poetry buildpack which also adds the dev
> dependencies export.
>
> I hear about this bug for the first time, but if it still persists I could
> and would like to add --without-hashes as to the default args.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/python-poetry/poetry/pull/2511#issuecomment-722230340>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHPJAGF5FZPCVNYMCXR5JLSOJQH7ANCNFSM4NV6E2YA>
> .
>

Cielquan commented:
> I didn't know about the argument override, that's cool! In that case it shouldn't be necessary to add the dev dependencies feature (though it's maybe worth documenting). ++++1 to this PR from me :)
> […](#)
> On Thu, Nov 5, 2020 at 12:40 AM Christian Riedel ***@***.***> wrote: Can you add a hook to also export dev dependencies? (poetry export --dev -f requirements.txt -o requirements-test.txt). This is useful for CI environments like Heroku <https://devcenter.heroku.com/articles/heroku-ci#python>. - id: poetry-export-dev name: poetry-export-dev description: run poetry export --dev to sync lock file with requirements-test.txt entry: poetry export language: python language_version: python3 pass_filenames: false args: ["--dev", "-f", "requirements.txt", "-o", "requirements-test.txt"] (I'm not sure if "dev" -> "test" is the ideal mapping, but poetry doesn't have a "test" section, and things like pytest are usually installed under dev in my experience) I would not like to do that as it would be only doubling the existing poetry-export hook with slightly other default args. I don't know how well you know pre-commit but the values in .pre-commit-hooks.yaml are only default values which you can overwrite/customize in your own .pre-commitconfig.yaml. So my recommended and preferred way of solving this would be to customize the hook in your .pre-commit-config.yaml like so: repos: - repo: https://github.com/python-poetry/poetry rev: <branch name / tag / sha> hooks: - id: poetry-export name: poetry-export-dev args: ["--dev", "-f", "requirements.txt", "-o", "requirements.txt"] For my example you have to add a corresponding rev value after this PR is merged. You just need to overwrite the default args and optionally you can overwrite the name, which is shown when pre-commit runs, too. ------------------------------ Also in the poetry Heroku buildpack <[https://github.com/StriveTalent/python-poetry-buildpack/blob/5b9d3e800467857d7d2670d2c2b0f66cbde2eb0e/bin/compile#L44-L50>[1]](https://github.com/StriveTalent/python-poetry-buildpack/blob/5b9d3e800467857d7d2670d2c2b0f66cbde2eb0e/bin/compile#L44-L50>%5B1%5D) and in this comment <[#2457 (comment)](https://github.com/python-poetry/poetry/issues/2457#issuecomment-713131930)>, the file is exported with the --without-hashes option. This appears to be due to this issue: [pypa/pip#4995](https://github.com/pypa/pip/issues/4995) <[pypa/pip#4995](https://github.com/pypa/pip/issues/4995)>, which relates to editable dependencies. It seems to me that since the bug where Poetry was exporting local dependencies as editable has been fixed, the --without-hashes argument shouldn't be needed, but wanted to confirm that here? 1: This is our fork of the poetry buildpack which also adds the dev dependencies export. I hear about this bug for the first time, but if it still persists I could and would like to add --without-hashes as to the default args. — You are receiving this because you commented. Reply to this email directly, view it on GitHub <[#2511 (comment)](https://github.com/python-poetry/poetry/pull/2511#issuecomment-722230340)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAHPJAGF5FZPCVNYMCXR5JLSOJQH7ANCNFSM4NV6E2YA> .

The documentation part is a good point. I will revisit die doc-changes and add a sentence or two.

s0undt3ch commented:
Will this ever go in?

teauxfu commented:
This seems pretty cool to me, and useful also, even if it were only to be included in the docs in some way

Cielquan commented:
I updated the branch to fix the merge conflict.

@python-poetry/triage Now waiting for workflow and PR approval.

Also for @s0undt3ch and @teauxfu and other people wanting to use this pre-merge: please see https://github.com/python-poetry/poetry/issues/2457#issuecomment-737297790 as temporary solution

rasulkireev commented:
Spent at least 3 hours trying to get this functionality, before finding this PR. 

@Cielquan Thanks for doing the work on this PR! Hopefully the team approves it soon.

graingert commented on .pre-commit-hooks.yaml:
```diff
@@ -0,0 +1,24 @@
+  entry: poetry check
+  language: python
+  language_version: python3
+  pass_filenames: false
```
I think you need
```suggestion
  pass_filenames: false
  require_serial: true
  files: ^*/(pyproject\.toml|poetry.lock)$
```

on all of these

Cielquan replied:
Why do you think I need this?

Line 8: By default this setting is `false` but why switching it? The hook calls `poetry check` which is one process in this context anyway. So I see no improvement here.

Line 9: `poetry check` does not take arguments like file names. Therefore this results in an `Too many arguments.` error. Same reason why `pass_filenames: false` is set.

I also run the exact same config from this PR in my [temp. mirror repo](https://github.com/Cielquan/mirrors-poetry) and it just works fine.

graingert replied:
without this I get:

```
poetry-check.........................................(no files to check)Skipped
poetry-lock..........................................(no files to check)Skipped
poetry-export........................................(no files to check)Skipped
```

Cielquan replied:
For me its working just fine.

relevant `.pre-commit-config.yaml` part:
```yaml
repos:
  - repo: https://github.com/Cielquan/mirrors-poetry
    rev: 87bf4eb48cd53dc409165358a84b1687b03efc9c  # frozen: 1.2.0a1
    hooks:
      - id: poetry-check
      - id: poetry-lock
      - id: poetry-export
```
Output:
```console
poetry-check.............................................................Passed
poetry-lock..............................................................Failed
- hook id: poetry-lock
- files were modified by this hook

Updating dependencies
Resolving dependencies... (15.7s)

Writing lock file

poetry-export............................................................Passed
```

I really do not understand your output, because if you use "my config", `pre-commit` does not send filenames to the hook and therefore does not require any and therefore should never show `(no files to check)Skipped`. When I delete my `pyproject.toml` file I get the error below for all 3 hooks. And this error comes from the hook and not `pre-commit`.
```console
poetry-check.............................................................Failed
- hook id: poetry-check
- exit code: 1

RuntimeError

  Poetry could not find a pyproject.toml file in /home/krys/Projects/.testing/python_test-cielquan or its parents

  at ~/.cache/pre-commit/repoqhauindm/py_env-python3/lib/python3.8/site-packages/poetry/core/factory.py:395 in locate
      391│             if poetry_file.exists():
      392│                 return poetry_file
      393│ 
      394│         else:
    → 395│             raise RuntimeError(
      396│                 "Poetry could not find a pyproject.toml file in {} or its parents".format(
      397│                     cwd
      398│                 )
      399│             )

```

graingert replied:
Are you running with `pre-commit run --all-files`?

Cielquan replied:
Yes I did for the answer above but after reading you question I ran and got
```console
$ pre-commit install
$ touch tesfile
$ git add testfile
$ git commit -m "add testfile"
poetry-check.............................................................Passed
poetry-lock..............................................................Passed
poetry-export............................................................Passed
...
```

graingert replied:
yes, it always runs if any files change, making pre-commit *much* slower

asottile replied:
@graingert is correct here (though `require_serial` is implied by `pass_filenames: false`) -- it should specify `files:` so it doesn't run all the time

Cielquan replied:
Ah ok .. I get the reasoning now. Thanks.
I will add `files: '^.*/pyproject\.toml$'`  for `check` and `files: '^.*/poetry\.lock$'` for `export` hook.

graingert commented on .pre-commit-hooks.yaml:
```diff
@@ -0,0 +1,24 @@
+  language: python
+  language_version: python3
+  pass_filenames: false
+  args: ["-f", "requirements.txt", "-o", "requirements.txt"]
```
poetry-export makes a constraints.txt file, not a requirements.txt file

```suggestion
  args: ["-f", "constraints.txt", "-o", "constraints.txt"]
```

graingert replied:
ah the poetry export is no-longer a valid constraints file https://github.com/python-poetry/poetry/issues/3472#issuecomment-744356551

Cielquan replied:
I see more use for/of `requirements.txt` files (correct me if I am wrong here) and these settings are only default which one can locally overwrite if one wants or needs to,

graingert commented on .pre-commit-hooks.yaml:
```diff
@@ -0,0 +1,24 @@
+  entry: poetry lock
+  language: python
+  language_version: python3
+  pass_filenames: false
```
I think it makes sense to default to "no-update" here - otherwise running this on CI for the same commit will break when new packages are uploaded to pypi.org

```suggestion
  pass_filenames: false
  args: ["--no-update"]
```

Cielquan replied:
I run `pre-commit` via GH Actions and also use `pre-commit.ci` and I do not think this is a good idea.

- You run `pre-commit` much more often locally than in CI. Therefore I prefer defaults for this case.
- My intention for this hook is to keep the lock file up-to-date and `--no-update` partially works against that.
- Also it does not happen that often, that a package is released while `pre-commit` is running in CI. When the run fails you can update the dependency accordingly and push again to trigger a new run. Or wait for e.g. `dependabot`.

Workarounds:
- You can run `SKIP=poetry-lock pre-commit run --all-files` and `poetry lock --no-update` in CI.
- `pre-commit.ci` can not only lint/check the files, but also commit changes from e.g. a formater back to a PR.

pkoch commented:
What would it take to get this in?

pkoch commented:
Poking @sdispater and @abn, since they seem like the most active maintainers.

pkoch commented:
Meanwhile, made this: https://github.com/pkoch/mirror-poetry

Cielquan commented:
> Meanwhile, made this: https://github.com/pkoch/mirrors-poetry

Already made this a while ago: https://github.com/Cielquan/mirrors-poetry
Also have it linked in my comment above.

pkoch commented:
Ah, neat! Thank you! I failed to notice that. That's an hour of my file i don't get back. :/

Cielquan commented:
Updated the PR to be ready for merging after approval.

neersighted commented on docs/pre-commit-hooks.md:
```diff
@@ -0,0 +1,101 @@
+
+## Usage
+
+For how to use `pre-commit` please see the [official documentation](https://pre-commit.com/).
```
```suggestion
For more information on how to use pre-commit please see the [official documentation](https://pre-commit.com/).
```

neersighted commented on docs/pre-commit-hooks.md:
```diff
@@ -0,0 +1,101 @@
+
+For how to use `pre-commit` please see the [official documentation](https://pre-commit.com/).
+
+Example `.pre-commit-config.yaml` file:
```
```suggestion
A full `.pre-commit-config.yaml` example:
```

neersighted commented on docs/pre-commit-hooks.md:
````diff
@@ -0,0 +1,101 @@
+    verbose: true
+```
+
+Or to put the `dev` dependencies into the `requirements.txt` also use this:
````
```suggestion
Also, `--dev` can be added to `arguments` to write `dev` dependencies to `requirements.txt`:
```

(Does this account for groups and the deprecation of `--dev`? It might need an update for the new syntax/paradigm.)

Cielquan replied:
It currently does not. I will look it up and fix it.

Cielquan replied:
Well the official docs (on master) state `--dev` as an option for the `export` command.

https://python-poetry.org/docs/master/cli/#options-11

neersighted replied:
LGTM then.

neersighted commented on docs/pre-commit-hooks.md:
```diff
@@ -0,0 +1,101 @@
+`args: ["-f", "requirements.txt", "-o", "requirements.txt"]` are the default arguments
+which will create/update the requirements.txt file in the current working directory.
+
+For output to the console change the arguments and add `verbose: true` to `poetry-export`
+in your `.pre-commit-config.yaml` file like so:
```
```suggestion
For output to the console change the arguments and add `verbose: true` to `poetry-export`
in your `.pre-commit-config.yaml` file like so:
```
```suggestion
You may add `verbose: true` in your `.pre-commit-config.yaml` in order to output to the
console:
```

neersighted commented on docs/pre-commit-hooks.md:
```diff
@@ -0,0 +1,101 @@
+The hook takes the same arguments as the poetry command.
+For more information see the [export command](/docs/cli#export).
+
+`args: ["-f", "requirements.txt", "-o", "requirements.txt"]` are the default arguments
+which will create/update the requirements.txt file in the current working directory.
```
```suggestion
The default arguments are `args: ["-f", "requirements.txt", "-o", "requirements.txt"]`,
which will create/update the requirements.txt file in the current working directory.
```

neersighted commented on docs/pre-commit-hooks.md:
```diff
@@ -0,0 +1,101 @@
+
+
+{{% note %}}
+If you specify the `args:` section for a hook in your pre-commit config
+the default `args:` are overwritten. So if you want to add arguments
+you need to specify the default ones in your config also.
```
```suggestion
If you specify the `args:` for a hook in your `.pre-commit-config.yaml`,
the defaults are overwritten. You must fully specify all arguments for
your hook if you make use of `args:`.
```

neersighted commented on docs/pre-commit-hooks.md:
```diff
@@ -0,0 +1,101 @@
+
+# pre-commit hooks
+
+`pre-commit` is a tool which manages running of
+[git hook](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks) scripts.
+See the official documentation for more information: [pre-commit.com](https://pre-commit.com/)
```
Sorry about all these nitpicks -- there's lots of passive voice and odd grammar that I'd like to correct before merge.

```suggestion
pre-commit is a framework for building and running
[git hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks).
See the official documentation for more information: [pre-commit.com](https://pre-commit.com/)
```

neersighted commented on docs/pre-commit-hooks.md:
```diff
@@ -0,0 +1,101 @@
+[git hook](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks) scripts.
+See the official documentation for more information: [pre-commit.com](https://pre-commit.com/)
+
+Below you can see the available `pre-commit` hooks you can use.
```
```suggestion
This document provides a list of available pre-commit hooks provided by Poetry.
```

neersighted commented on docs/cli.md:
```diff
@@ -529,6 +529,10 @@ As such, `exit` should be used to properly exit the shell and the virtual enviro
 and returns a detailed report if there are any errors.
 
+{{% note %}}
+This command is also available as a `pre-commit` hook. See [pre-commit hooks](/docs/pre-commit-hooks#poetry-check) for more information.
```
```suggestion
This command is also available as a pre-commit hook. See [pre-commit hooks](/docs/pre-commit-hooks#poetry-check) for more information.
```

neersighted commented on docs/cli.md:
```diff
@@ -547,6 +551,7 @@ This command locks (without installing) the dependencies specified in `pyproject
 
 {{% note %}}
 By default, this will lock all dependencies to the latest available compatible versions. To only refresh the lock file, use the `--no-update` option.
+This command is also available as a `pre-commit` hook. See [pre-commit hooks](/docs/pre-commit-hooks#poetry-lock) for more information.
```
```suggestion
This command is also available as a pre-commit hook. See [pre-commit hooks](/docs/pre-commit-hooks#poetry-lock) for more information.
```

neersighted commented on docs/cli.md:
```diff
@@ -595,6 +600,7 @@ poetry export -f requirements.txt --output requirements.txt
 
 {{% note %}}
 Only the `requirements.txt` format is currently supported.
+This command is also available as a `pre-commit` hook. See [pre-commit hooks](/docs/pre-commit-hooks#poetry-export) for more information.
```
```suggestion
This command is also available as a pre-commit hook. See [pre-commit hooks](/docs/pre-commit-hooks#poetry-export) for more information.
```

neersighted commented:
Minor doc changes requested. I'd also love people to give the docs a close look (though I don't necessarily feel a need to hold up merging this) -- bad docs for pre-commit hooks are super common and super annoying in the ecosystem.

Cielquan commented:
> Minor doc changes requested. I'd also love people to give the docs a close look (though I don't necessarily feel a need to hold up merging this) -- bad docs for pre-commit hooks are super common and super annoying in the ecosystem.

Thanks for the review and suggestions. Your changed phrasing is easier to read than mine. 👍🏾 
Now only the issue with the docs for groups is missing above.

Cielquan commented:
Thanks for merging 😃

MaxWinterstein commented:
Would love to use this at [my repo](https://github.com/MaxWinterstein/toogoodtogo-ha-mqtt-bridge/pull/31), but it seems there is a problem at `pre-commit.ci` side I guess?

I can't find an already cached version, and the installation always fails on me, see e.g.:

https://results.pre-commit.ci/run/github/327882738/1639672185.PASU-kRYTIufuGkEgXVAAA

graingert commented:
> Would love to use this at [my repo](https://github.com/MaxWinterstein/toogoodtogo-ha-mqtt-bridge/pull/31), but it seems there is a problem at `pre-commit.ci` side I guess?
> 
> I can't find an already cached version, and the installation always fails on me, see e.g.:
> 
> https://results.pre-commit.ci/run/github/327882738/1639672185.PASU-kRYTIufuGkEgXVAAA

You need a paid pre-commit.ci account to access the network during the running of hooks

MaxWinterstein commented:
> You need a paid pre-commit.ci account to access the network during the running of hooks

Oh, I see, found some notes about this here: https://github.com/pre-commit-ci/issues/issues/55#issuecomment-822481997

So, at GitHub level, using pre-commit.ci, this hook is unusable.

This might be worth mentioning in the documents?

floatingpurr commented:
I'm trying the `poetry-export` hook but it doesn't work. I need to override the `files` value this way:

```yaml
  # export python requirements
  - repo: https://github.com/python-poetry/poetry
    rev: 2fa6c17 # put here the next release with the hooks
    hooks:
      - id: poetry-export
        files: '^.*poetry\.lock$'
```

with the default value (i.e., `'^.*/poetry\.lock$'`) it didn't work. I'm afraid that the same holds for `poetry-check`. If necessary, I can open up a PR about it.

vwxyzjn commented:
For those of you who stumbled across this post like I do, you can do:

```
  - repo: https://github.com/python-poetry/poetry
    rev: 1.2.0b1
    hooks:
      - id: poetry-export
        args: ["-f", "requirements.txt", "-o", "requirements.txt"]
```
